File: src/moleminer/sources/hackernews.py
"""Hacker News search via Algolia API (free, no key needed)."""

from __future__ import annotations

from urllib.parse import quote_plus

from moleminer.models import SearchResult
from moleminer.sources.base import BaseSource
from moleminer.utils.http import fetch_json

_HN_API = "https://hn.algolia.com/api/v1/search"
_HN_ITEM_URL = "https://news.ycombinator.com/item?id={}"


class HackerNewsSource(BaseSource):
    name = "hackernews"
    source_type = "api"
    requires_auth = False

    def enabled(self) -> bool:
        return True

    async def search(self, queries: list[str]) -> list[SearchResult]:
        results: list[SearchResult] = []
        for query in queries:
            hits = await self._search_one(query)
            results.extend(hits)
        return results

    async def _search_one(self, query: str) -> list[SearchResult]:
        url = f"{_HN_API}?query={quote_plus(query)}&tags=story&hitsPerPage=20"
        data = await fetch_json(url)
        results: list[SearchResult] = []
        for hit in data.get("hits", []):
            item_url = hit.get("url") or _HN_ITEM_URL.format(hit["objectID"])
            results.append(
                SearchResult(
                    title=hit.get("title", ""),
                    url=item_url,
                    source="hackernews",
                    snippet=hit.get("title", ""),
                    result_type="lead",
                    timestamp=hit.get("created_at"),
                    metadata={
                        "points": hit.get("points", 0),
                        "num_comments": hit.get("num_comments", 0),
                        "hn_id": hit.get("objectID"),
                        "author": hit.get("author"),
                    },
                )
            )
        return results

File: ts/src/sources/hackernews.ts
/** Hacker News search via Algolia API (free, no key needed). */

import type { Config } from '../config.js';
import type { SearchResult } from '../models.js';
import { BaseSource } from './base.js';
import { fetchJson } from '../utils/http.js';

const HN_API = 'https://hn.algolia.com/api/v1/search';
const HN_ITEM_URL = 'https://news.ycombinator.com/item?id=';

interface HNHit {
  objectID: string;
  title?: string;
  url?: string;
  created_at?: string;
  points?: number;
  num_comments?: number;
  author?: string;
}

interface HNResponse {
  hits: HNHit[];
}

export class HackerNewsSource extends BaseSource {
  readonly name = 'hackernews';
  readonly sourceType = 'api' as const;
  readonly requiresAuth = false;

  enabled(_config: Config): boolean {
    return true;
  }

  protected override async searchOne(query: string): Promise<SearchResult[]> {
    const params = new URLSearchParams({
      query,
      tags: 'story',
      hitsPerPage: '20',
    });
    const data = await fetchJson<HNResponse>(`${HN_API}?${params}`);

    return (data.hits ?? []).map((hit) => ({
      title: hit.title ?? '',
      url: hit.url || `${HN_ITEM_URL}${hit.objectID}`,
      source: 'hackernews',
      snippet: hit.title ?? '',
      resultType: 'lead' as const,
      timestamp: hit.created_at,
      metadata: {
        points: hit.points ?? 0,
        num_comments: hit.num_comments ?? 0,
        hn_id: hit.objectID,
        author: hit.author,
      },
    }));
  }
}
